llvm/lib/Transforms/Scalar/StraightLineStrengthReduce.cpp

```diff
@@ -269,17 +616,289 @@ FunctionPass *llvm::createStraightLineStrengthReducePass() {
+        return false;
+
+      // Record delta if none has been found yet, or the new delta is
+      // a constant that is better than the existing delta.
+      if (!C.Delta || isa<ConstantInt>(AvailableVal)) {
+        C.Delta = AvailableVal;
+        C.Basis = Basis;
+        C.DeltaKind = K;
+      }
+      return isa<ConstantInt>(C.Delta);
+    };
+  };
+
+  if (const auto *BaseDeltaCandidates =
+          CandidateDict.getCandidatesWithDeltaKind(C, Candidate::BaseDelta)) {
+    if (SearchFrom(*BaseDeltaCandidates, For(Candidate::BaseDelta))) {
+      LLVM_DEBUG(dbgs() << "Found delta from Base: " << *C.Delta << "\n");
+      return;
+    }
+  }
+
+  if (const auto *StrideDeltaCandidates =
+          CandidateDict.getCandidatesWithDeltaKind(C, Candidate::StrideDelta)) {
+    if (SearchFrom(*StrideDeltaCandidates, For(Candidate::StrideDelta))) {
+      LLVM_DEBUG(dbgs() << "Found delta from Stride: " << *C.Delta << "\n");
+      return;
+    }
+  }
+
+  // If we did not find a constant delta, we might have found a variable delta
+  if (C.Delta) {
+    LLVM_DEBUG({
+      dbgs() << "Found delta from ";
+      if (C.DeltaKind == Candidate::BaseDelta)
+        dbgs() << "Base: ";
+      else
+        dbgs() << "Stride: ";
+      dbgs() << *C.Delta << "\n";
+    });
+    assert(C.DeltaKind != Candidate::InvalidDelta && C.Basis);
+  }
+}
+
+// Compress the path from `Basis` to the deepest Basis in the Basis chain
+// to avoid non-profitable data dependency and improve ILP.
+// X = A + 1
+// Y = X + 1
+// Z = Y + 1
+// ->
+// X = A + 1
+// Y = A + 2
+// Z = A + 3
+// Return the delta info for C aginst the new Basis
+auto StraightLineStrengthReduce::compressPath(Candidate &C,
+                                              Candidate *Basis) const
+    -> DeltaInfo {
+  if (!Basis || !Basis->Basis || C.CandidateKind == Candidate::Mul)
+    return {};
+  Candidate *Root = Basis;
+  Value *NewDelta = nullptr;
+  auto NewKind = Candidate::InvalidDelta;
+
+  while (Root->Basis) {
+    Candidate *NextRoot = Root->Basis;
+    if (C.Base == NextRoot->Base && C.StrideSCEV == NextRoot->StrideSCEV &&
+        isSimilar(C, *NextRoot, Candidate::IndexDelta)) {
+      ConstantInt *CI = cast<ConstantInt>(getIndexDelta(C, *NextRoot));
+      if (CI->isZero() || CI->isOne() || isa<SCEVConstant>(C.StrideSCEV)) {
+        Root = NextRoot;
+        NewKind = Candidate::IndexDelta;
+        NewDelta = CI;
+        continue;
+      }
+    }
+
+    const SCEV *CandPart = nullptr;
+    const SCEV *BasisPart = nullptr;
+    auto CurrKind = Candidate::InvalidDelta;
+    if (C.Base == NextRoot->Base && C.Index == NextRoot->Index) {
+      CandPart = C.StrideSCEV;
+      BasisPart = NextRoot->StrideSCEV;
+      CurrKind = Candidate::StrideDelta;
+    } else if (C.StrideSCEV == NextRoot->StrideSCEV &&
+               C.Index == NextRoot->Index) {
+      CandPart = C.Base;
+      BasisPart = NextRoot->Base;
+      CurrKind = Candidate::BaseDelta;
+    } else
+      break;
+
```

Comment:
It's not clear whether `CondPart` and `BasisPart` are expected to always be non-null past this point or not.

If we expect them to always be non-null, I'd add an appropriate assert here.